### DISCUSSION TITLE
Unable to reach remote machines from Kubernetes pods using strongswan IPsec tunnel

### DISCUSSION THREAD

**Discussion by aleemgsl** [strongswan/strongswan#1442] [Q&A] [answered]
Hi All,

We have deployed a containerized strongswan using "vimagick/strongswan" image. Our requirement is as below:

1. On a baremetal server, we have multiple pods running (all are in the same network). Strongswan is one of them.
2. We would like to access one of the machines present on remote side (lets say X.X.X.X). This Machine should be accessible from all the pods. 
3. The pods on local side should also be accessible from remote side.

Problem We are facing:

1. We are able to access the X.X.X.X machine from strongswan pod. But we can not ping X.X.X.X machine from any other pods.

We have added routes on local pods to send traffic to X.X.X.X via strongswan pod.

Below is the config file:

#
config setup
        #charondebug="ike 3, knl 3, cfg 3, chd 3, dmn 3"
        charondebug="dmn 5, mgr 5, ike 5, chd 5, job 5, enc 4, knl 2, enc 5, net 2, asn 2, lib 5, esp 5, tls 2, tnc 2, imc 2, imv 2, pts 2, cfg 5"
conn "ikev2"
        auto=start
        keyexchange=ikev2
        ike=aes256-sha2_256-modp2048
        esp=aes256-sha2_256
        type=tunnel
        leftsourceip=%modeconfig
        leftcert=abc.pem
        right=<Public_IP_of_remote_side>
        rightid=%any
        rightsubnet=X.X.X.0/26
        authby=rsasig
        dpddelay=5
        dpdtimeout=20
        dpdaction=restart
        closeaction=restart
        keyingtries=%forever
		
		
Diagram:	

![image](https://user-images.githubusercontent.com/110398960/204515589-880fc2ed-d321-4411-ad04-9f072d528dbd.png)

**Comment by tobiasbrunner** [comment, marked as answer]
Check the negotiated traffic selectors. Since you don't have a net-to-net config and request a virtual IP, you'll see that the local traffic selector is a single address (the virtual IP address assigned by the remote router). Traffic from your other pods won't match the resulting IPsec policies. To achieve that you have to NAT that traffic to the virtual IP (since it's probably dynamically assigned via an updown script). Alternatively, change the config so that you actually negotiate a site-to-site policy for all IPs your pods have.

**Reply by aleemgsl** [reply]
Thanks for the reply. 

Can you please share the sample config to negotiate site-to-site policy for all the IPs?

**Reply by tobiasbrunner** [reply]
Just remove `leftsourceip` and add `leftsubnet=<subnet for pods>/<mask>`. The remote end has to be configured accordingly.

**Reply by aleemgsl** [reply]
Hi @tobiasbrunner

As suggested, we made changes by adding leftsubnet=192.168.84.0/24 (Pod network). We could see the association post ipsec restart as:

ikev2{1}: 192.168.84.0/24 === X.X.X.0/26

Though, we still do not see the successful ping from pods to the remote machine.

Can you please let us know how to force/send traffic going out from pod towards X.X.X.0/26 via IPsec tunnel?

We tried adding route on host machine to send traffic to X.X.X.0/26 via virtual IP (possible only when we keep leftsourceip=%modeconfig and leftsubnet).
With only leftsubnet=192.168.84.0/24, we can not add this route.

The current config looks like the below:

config setup
#charondebug="ike 3, knl 3, cfg 3, chd 3, dmn 3"
charondebug="dmn 5, mgr 5, ike 5, chd 5, job 5, enc 4, knl 2, enc 5, net 2, asn 2, lib 5, esp 5, tls 2, tnc 2, imc 2, imv 2, pts 2, cfg 5"
conn "ikev2"
auto=start
keyexchange=ikev2
ike=aes256-sha2_256-modp2048
esp=aes256-sha2_256
type=tunnel
leftcert=abc.pem
leftfirewall=yes
leftsubnet=192.168.84.0/24
right=<Public_IP_of_remote_side>
rightid=%any
rightsubnet=X.X.X.0/26
authby=rsasig
dpddelay=5
dpdtimeout=20
dpdaction=restart
closeaction=restart
keyingtries=%forever

Also tried by adding "leftsourceip=%modeconfig" to the above config to add a route to send traffic to the virtual IP.

We tried both containerized (adding a route on pods to send X.X.X.0/26 traffic to strongswan container) as well as non-containerized strongswan deployment with above configs.

**Reply by tobiasbrunner** [reply]
Please see [here](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html).

**Reply by aleemgsl** [reply]
Hi,

I have deployed strongswan on two new machines using below configs:

Machine 1 : 10.x.x.112 (Docker Network: 192.168.84.0/24)

config setup
        charondebug="all"
        uniqueids=yes
        strictcrlpolicy=no
conn edge-to-intelcord-centralsite
        authby=secret
        type=tunnel
        auto=start
        keyexchange=ikev2
        authby=secret
        left=10.x.x.112
        leftid=10.x.x.112
        leftsubnet=192.168.84.0/24
        right=10.x.x.190
        rightsubnet=10.20.84.0/24
        ike=aes256-sha1-modp1024!
        esp=aes256-sha1!
        aggressive=no
        keyingtries=%forever
        ikelifetime=28800s
        lifetime=3600s
        dpddelay=30s
        dpdtimeout=120s
        dpdaction=restart
        auto=start
		
		
Machine 2: 10.x.x.190 (Docker Network: 10.10.84.0/24)	

config setup
        charondebug="all"
        uniqueids=yes
        strictcrlpolicy=no
conn omec_new_centralsite-to-edge
        authby=secret
        type=tunnel
        auto=start
        keyexchange=ikev2
        authby=secret
        left=10.x.x.190
        leftid=10.x.x.190
        leftsubnet=10.20.84.0/24
        right=10.x.x.112
        rightsubnet=192.168.84.0/24
        ike=aes256-sha1-modp1024!
        esp=aes256-sha1!
        aggressive=no
        keyingtries=%forever
        ikelifetime=28800s
        lifetime=3600s
        dpddelay=30s
        dpdtimeout=120s
        dpdaction=restart
        auto=start
		
		
10.20.84.0/24 === 192.168.84.0/24
		
IPSEC tunnel is up and I can reach the remote machines from the local pods and vise versa. But when I capture pcap to analyze the packets, I see it plain traffic (plain icmp).

In table 220 I see a route added as below :

root@GS-3378:~# ip route list table 220
192.168.84.0/24 via 10.x.x.112 dev enp0s31f6 proto static src 10.20.84.0

Can you please let us know why we see plain ICMP packets and not esp packets?

**Reply by tobiasbrunner** [reply]
Docker installs NAT rules for its virtual networks. You have to exclude traffic that matches your IPsec policies from that (see the page I linked before).

**Reply by aleemgsl** [reply]
After disabling NATTING for the traffic going out of POD, it started to work.
